pennylane/transforms/optimization/commute_through_controls_targets.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transforms for pushing commuting gates through targets/control qubits."""
+
+from itertools import chain
+
+from pennylane import apply
+from pennylane.wires import Wires
+from pennylane.transforms import qfunc_transform
+
+from .optimization_utils import find_next_gate
+
+# Gates that commute among each particular basis
+commuting_gates = {
+    "X": ["PauliX", "RX", "SX"],
+    "Y": ["PauliY", "RY"],
+    "Z": ["PauliZ", "RZ", "PhaseShift", "S", "T"],
+}
```

Comment:
Out of curiosity: could we not have a `basis` property on gates instead? e.g.:

```python
class RX(Operation):
    basis = "X"
    ...

class CRX(Operation):
    basis = "X"
    ...
```

This could replace the dictionary here as well as the `target_basis` property?

Reply:
Alternatively, you could instead define the controlled target basis as a dictionary of gates here?

Reply:
Not really suggesting one over the other, I think both approaches have their place. Just gathering my thoughts: I guess I'm trying to determine why `commuting_gates` is transform-specific data, while `target_basis` is a Operation-specific data.

Is `commuting_gates` explicitly linked to the compilation transform?

Reply:
Awesome, I like the idea of specifying a basis; then we can just check that they match, and there is no need to have any of this info here. Nothing is compilation-specific, the gates either commute or don't, it's just we aren't right now making use of that information anywhere else.